### DISCUSSION TITLE
How to find out which column in a query is the cause of the `UnexpectedNullError`?

### DISCUSSION THREAD

**Discussion by davidatsurge** [diesel-rs/diesel#3503] [Q&A] [answered]
Title says it all.

Also, it's a bit surprising that I'd get this error at all, as I was hoping that an advantage of using `diesel` was to avoid such type mismatches at compile time. Are there other such known corner-cases where the "if-it-compiles-it-works" promise doesn't quite uphold? I'm using Postgres, if that's relevant.

**Comment by weiznich** [comment, marked as answer]
Its impossible to help you here without knowing quite a few more details. Generally speaking: Assuming you don't use a custom query dsl extension and you did not change your schema.rs file and you did not call one of the methods that explicitly change the nullability of an expression and that's not caused by `sql_query` I would consider that a bug. Given the current report, there is not much we can do to track this down. You would need to provide an reproducible example for further help.

**Reply by davidatsurge** [reply]
So I think I narrowed it down + came up with a repro. [Here's the repro](https://github.com/davidatsurge/left_outer_join_array_agg_repro).

Here's an even more minimal summary of that repro:
```rust
use diesel::{prelude::*, sql_types::SingleValue};
use std::error::Error;

mod schema {
    // @generated automatically by Diesel CLI.
    diesel::table! {
        descendants (id) {
            id -> Text,
            parent_id -> Nullable<Text>,
        }
    }

    diesel::table! {
        parents (id) {
            id -> Text,
        }
    }

    diesel::joinable!(descendants -> parents (parent_id));

    diesel::allow_tables_to_appear_in_same_query!(descendants, parents,);
}

// I _think_ this is correct.
sql_function! {
    #[aggregate]
    #[sql_name = "ARRAY_AGG"]
    fn array_agg<ST :SingleValue >(x: ST) -> Array<ST>;
}

fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
    use schema::*;

    let database_url = dotenv::var("DATABASE_URL")?;
    let mut conn = PgConnection::establish(&database_url)?;

    parents::table
        .left_join(descendants::table)
        // HERE: This is the cause of the UnexpectedNullError. I believe diesel should reject this
        // code. The correct way to perform an ARRAY_AGG over a nullable column should instead be
        // `.select(array_agg(descendants::id.nullable()))`, which diesel already accepts.
        .select(array_agg(descendants::id).nullable())
        .load::<Option<Vec<String>>>(&mut conn)?;

    Ok(())
}
```

**Reply by weiznich** [reply]
Thanks for providing this minimal example. This makes it much clearer what is happening here. 

Lets have a look at the difference between `array_agg(descendants::id).nullable()` and `array_agg(descendants::id.nullable())`:

Given your definition of `array_agg`, the first expression generates a SQL expression of the type `Nullable<Array<Text>>` while the second one generates an expression of the type `Array<Nullable<Text>>`. I would argue that's because your definition of `array_agg`, that literally forwards the input SQL type. The big question is now: How can that function be written in a way that does not have this issue? I do not have an answer yet, but that's something that must be figured out until an `array_agg` function is provided by diesel itself.

**Reply by davidatsurge** [reply]
Thanks Georg!

Coming up with this minimal example took a bit of time because I had to narrow down which column in a big query was causing the issue. That is why I asked if there was a way to find out which column in a query is the cause of the `UnexpectedNullError`. Just in case similar things happen in the future, can I ask you that question again? Unless I missed something, I don't think you addressed it.

**Reply by davidatsurge** [reply]
For bystanders watching, I gotta say that I just realized the definition of `array_agg` above is incorrect in _another_ way (that  is orthogonal to the issue we're discussing). According to [postgres docs](https://www.postgresql.org/docs/9.6/functions-aggregate.html), "array_agg returns null rather than an empty array when there are no input rows." So the diesel definition should be something like:

```rust
sql_function! {
    #[aggregate]
    #[sql_name = "ARRAY_AGG"]
    fn array_agg<ST :SingleValue >(x: ST) -> Nullable<Array<ST>>;
}
```

Note that that definition also suffers from the the same issue that I discuss above (ie, will cause a `UnexpectedNullError` at runtime if called on a left-joined, originally-not-null, column that has NULL values after the join).

**Reply by weiznich** [reply]
> Just in case similar things happen in the future, can I ask you that question again? Unless I missed something, I don't think you addressed it.

I didn't directly address that, but that's currently not possible as that error does not contain any information at all.
